linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1298,11 +1298,11 @@ private void closeInner() {
       final EqualityComparer<TKey> comparer) {
     return new AbstractEnumerable<TSource>() {
       public Enumerator<TSource> enumerator() {
-        final Enumerable<TKey> innerLookup =
-            comparer == null
-                ? inner.select(innerKeySelector).distinct()
-                : inner.select(innerKeySelector).distinct(comparer);
-
+        // CALCITE-2909 Delay the computation of the innerLookup until the moment when we are sure
+        // that it will be really needed, i.e. when the first outer enumerator item is processed
+        final Enumerable<TKey> innerLookup = comparer == null
+                ? Linq4j.lazyEnumerable(() -> inner.select(innerKeySelector).distinct())
+                : Linq4j.lazyEnumerable(() -> inner.select(innerKeySelector).distinct(comparer));
```

Comment:
Rather than inventing `LazyEnumerable`, could you change the type of `innerLookup` from `Enumerable<TKey>` to `Supplier<Enumerable<TKey>>` or (better) `Supplier<Enumerator<TKey>>` and use Guava's `MemoizingSupplier`

The goal, after all, is that the hash table is computed 0 times if not needed, and 1 time if needed 1 or more times. A memoizing supplier achieves that very well.

I don't see anything in the code that requires `innerLookup` to be an Enumerable.

My concern with LazyEnumerable is that Enumerable is, by design, a lazy data structure. When you invent another layer of laziness, it will be difficult to figure out which of the 2 lazy layers to put new features. It is already confusing enough.

Reply:
Ok, I'll work on that approach, which would mean that we could abandon [CALCITE-2937](https://issues.apache.org/jira/browse/CALCITE-2937)

Reply:
Originally I was guided by the fact that it is nicer (from API perspective) to use `T` (which may be lazy) rather than explicitly define `Lazy<T>`.  Some examples:
1. [Supplier.memoize](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier)) from guava. 
2. [StreamSupport.stream(Supplier<Spliterator>)](https://docs.oracle.com/javase/8/docs/api/java/util/stream/StreamSupport.html#stream-java.util.function.Supplier-int-boolean-)
3. [Observable.defer](http://reactivex.io/RxJava/javadoc/rx/Observable.html#defer-rx.functions.Func0-)

I'm sorry if I have wasted your time @rubenada . This change looks fine. Pls rebase and I'll merge it after 1.19 release.

Reply:
No problem @asereda-gs , yours was an interesting approach too